src/rewards/voters.rs

```diff
@@ -30,17 +42,241 @@ pub fn calculate_stake<'address>(
+    active_addresses: &ActiveAddresses,
+    total_rewards: Rewards,
+) -> HashMap<&'address Address, Rewards> {
     stake_per_voter
         .iter()
-        .map(|(k, v)| (*k, fixed::types::U64F64::from_num(*v) / total_stake as u128))
+        .map(|(k, v)| {
+            // if it doesnt appear in the votes count, it means it did not vote
+            let reward = if active_addresses.contains(k) {
+                Rewards::from(*v) / Rewards::from(total_stake) * total_rewards
+            } else {
+                Rewards::ZERO
+            };
+            (*k, reward)
+        })
         .collect()
 }
 
-/// get the proportional reward from a share and total rewards amount
-pub fn reward_from_share(share: U64F64, total_reward: u64) -> fixed::types::U64F64 {
-    fixed::types::U64F64::from_num(total_reward) * share
+pub type VoteCount = HashMap<String, u64>;
+pub type ActiveAddresses = HashSet<Address>;
+
+fn active_addresses(
+    vote_count: VoteCount,
+    block0: &Block0Configuration,
+    threshold: u64,
+) -> ActiveAddresses {
+    let discrimination = block0.blockchain_configuration.discrimination;
+    vote_count
+        .into_iter()
+        .filter_map(|(account_hex, count)| {
+            if count >= threshold {
+                Some(
+                    account_hex_to_address(account_hex, discrimination)
+                        .expect("Valid hex encoded UnspecifiedAccountIdentifier"),
+                )
+            } else {
+                None
+            }
+        })
+        .collect()
+}
+
+fn account_hex_to_address(
+    account_hex: String,
+    discrimination: Discrimination,
+) -> Result<Address, hex::FromHexError> {
+    let mut buffer = [0u8; 32];
+    hex::decode_to_slice(account_hex, &mut buffer)?;
+    let identifier: UnspecifiedAccountIdentifier = UnspecifiedAccountIdentifier::from(buffer);
+    Ok(Address::from(chain_addr::Address(
+        discrimination,
+        Kind::Account(
+            identifier
+                .to_single_account()
+                .expect("Only single accounts are supported")
+                .into(),
+        ),
+    )))
+}
+
+fn rewards_to_mainnet_addresses(
+    rewards: HashMap<&'_ Address, Rewards>,
+    snapshot: Snapshot,
+) -> HashMap<MainnetRewardAddress, Rewards> {
+    let mut res = HashMap::new();
+    for (addr, reward) in rewards {
+        let registrations = snapshot.registrations_for_voting_key(
+            addr.as_ref()
+                .public_key()
+                .expect("non account address")
+                .clone(),
+        );
+        let total_stake = registrations
+            .iter()
+            .map(|reg| Rewards::from(u64::from(reg.stake)))
+            .sum::<Rewards>();
+        dbg!(total_stake);
+        dbg!(&registrations);
+
+        for reg in registrations {
+            *res.entry(reg.reward_addr).or_default() +=
+                reward * Rewards::from(u64::from(reg.stake)) / total_stake;
+        }
+    }
+
+    res
+}
```

Comment:
This can be further simplified with the following algorithm:

- Sort out inactive voter keys
- Get all registrations that are above threshold for these keys
- Run the rewards distribution logic on these registrations

This would reduce the number of calculations to be done compared to the current approach and thus would make the code less error-prone.

Reply:
mmm, it should be what I do now